napari/components/cursor_event.py

```diff
@@ -0,0 +1,45 @@
+from typing import Optional, Tuple
+
+from ..utils.events import ReadOnlyModel
+from ._viewer_constants import CursorEventType
+
+
+class CursorEvent(ReadOnlyModel):
+    """Cursor event object that gets passed to mouse callbacks.
+
+    Attributes
+    ----------
+    position : tuple
+        Position of the cursor in world coordinates.
+    data_position : tuple
+        Position of the cursor in data coordinates if the
+        cursor event has been passed to a specific layer.
+        None otherwise.
+    canvas_position : tuple
+        Position of the cursor in canvas.
+    delta : tuple or None
+        Mouse movement on last event. None if outside the
+        canvas.
+    type : str
+        Type of last cursor event. Must be one of
+        * mouse_move: A mouse move event
+        * mouse_press: A mouse press event
+        * mouse_release: A mouse release event
+        * mouse_wheel: A mouse wheel event
+    inverted : bool
+        Flag if cursor movement is inverted.
+    is_dragging : bool
+        Flag if cursor is currently dragging.
+    modifiers : tuple
+        Identities of any modifier keys held during cursor event.
+    """
+
+    # fields
+    position: Tuple[float, ...] = (1, 1)
+    data_position: Optional[Tuple[float, ...]] = None
+    canvas_position: Tuple[int, int] = (0, 0)
+    delta: Tuple[float, float] = (0, 0)
```

Comment:
do you want to make those a named tuple  with x and y fields ?

Reply:
also you need to be using 2 convention `[x, x]` and `[x, ...]`, I guess that's because position and data_position can be 3 tuples ?

Reply:
could do, as you noted the full positions can actually be in nD (they take the position of the sliders into account to consider the full location of the pixel being hovered over)